### ISSUE TITLE
bumble-hci-bridge causing hci devices to dissapear?

### ISSUE THREAD

**Issue by mscottwvc** [google/bumble#78]
Running ubuntu 22.04.1
I run 
`bumble-hci-bridge android-emulator:mode=controller usb:0`
Which works as expected, however after I end this process (I am killing it, maybe there is a graceful way to shut this down?) the hci device I was using "hci0" in this case no longer shows up with a bluetooth command like hciconfig for example, it also no longer shows up if I use the Ubuntu GUI bluetooth.  The usb device still shows up with lsusb and dmesg does not show the hci device being removed.

**Comment by barbibulle** [comment]
Killing the process should be fine.
A couple things to check:
1/ What does `hciconfig` show before and after you start the bridge app?
2/ can you still use the `usb:0` device with Bumble after you killed the bridge process? (ex: `bumble-controller-info usb:0`)

If BlueZ is running, it is normal that it may try to re-attach to the dongle device when you unclaim it. See https://google.github.io/bumble/platforms/linux.html for some info on how to deal with that, and how to use `hciconfig <hciX> up` and   `hciconfig <hciX> down` to control that effect.

**Comment by mscottwvc** [comment]
1) Before running:
`hciconfig
hci0:	Type: Primary  Bus: USB
	BD Address: E4:FD:45:F5:70:84  ACL MTU: 1021:4  SCO MTU: 96:6
	DOWN 
	RX bytes:20237 acl:0 sco:0 events:3288 errors:0
	TX bytes:813282 acl:0 sco:0 commands:3286 errors:0
`
While running and after running, hciconfig can no longer see the device: 
`hciconfig`
2) bumble-controller-info works, the hci-bridge will also work again along with bumble-usb-probe.  Just BlueZ will not work anymore.  I am familiar with the use of hciconfig, but it will no longer show devices or turn on / off devices.  I also have some BlueZ C++ programs that I have been developing which also refuse to work in this state.  Unplugging and replugging the bluetooth device restores BlueZ connectivity.

**Comment by barbibulle** [comment]
This is puzzling... I've tried to replicate on a laptop running Ubuntu 22.04.1, and a BT851 USB dongle. I can't make `hci0` disappear.
Maybe something goes wrong when BlueZ tries to re-claim the dongle? (which might not show up in `dmesg` but may show up in BlueZ logs, possibly `btmon`). 
Also, which USB BT controller are you using? It is possible that some controllers get in a bad state when the USB interface is just closed without an explicit HCI reset before, and somehow that bad state causes issues with BlueZ but not Bumble?
Maybe someone else with a similar setup can try and replicate the issue? @tavip?

**Comment by mscottwvc** [comment]
Tested with two different bluetooth devices, Intel AX200 (internal usb) and realtek rtl8761b (external usb), both have same results.

**Comment by mscottwvc** [comment]
@barbibulle what version of Bluez do you have?

**Comment by barbibulle** [comment]
I did find out what's going on here. Somehow, I forgot that the `usb` transport module for Bumble, by default, asks to detach the kernel driver for a USB interface if there's one. `libusb` doesn't automatically re-attach the driver when the interface is released. So I pushed a small PR, #84, to use `libusb`'s auto-detach feature, which will automatically detach the kernel driver when the interface is claimed, and re-attach it when the interface is released. This seems to work well on my Linux machine.
